### PR TITLE
Feat/confusion matrix metric

### PR DESCRIPTION
**Pull request by JSabadin**
# Confusion Matrix Metric

In this PR, we have added the Confusion Matrix (CM) metric, which is logged using the `log_matrix` function from the `luxonis_ml` tracker. The Confusion Matrix is supported for the following tasks:

- Detection
- Classification
- Segmentation

We should merge this after [PR #209](https://github.com/luxonis/luxonis-ml/pull/209) in luxonis-ml is merged, as it enables the logging of a matrix.

### CODE REVIEW

File: luxonis_train/attached_modules/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                    self.n_classes, self.n_classes, dtype=torch.int64
+                ),
+                dist_reduce_fx="sum",
+            )
+        if self.is_detection:
+            self.add_state(
+                "detection_cm",
+                default=torch.zeros(
+                    self.n_classes + 1, self.n_classes + 1, dtype=torch.int64
+                ),  # +1 for background
```
[the remark this answers is not in the document]
Comment (JSabadin) [reply luxonis_train/attached_modules/metrics/confusion_matrix.py:92]: Yes, it is necessary. For example, look at how they compute it [here](https://github.com/ultralytics/yolov5/blob/master/utils/metrics.py#L124).

File: luxonis_train/attached_modules/metrics/confusion_matrix.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+from torch import Tensor
+from torchvision.ops import box_convert, box_iou
+
+from luxonis_train.enums import TaskType
+from luxonis_train.utils import Labels, Packet
+
+from .base_metric import BaseMetric
+
+
+class ConfusionMatrix(BaseMetric[Tensor, Tensor]):
```
[the remark this answers is not in the document]
Comment (JSabadin) [reply luxonis_train/attached_modules/metrics/confusion_matrix.py:13]: Yes, it works for bounding boxes in that case. The result is a 2x2 matrix of TP, FP, FN, and TN for person keypoint detection.

File: luxonis_train/attached_modules/metrics/confusion_matrix.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+            tasks.
+        @type targets: dict[str, Tensor]
+        @param targets: A dictionary containing targets for all tasks.
+        """
+        if "classification" in predictions and "classification" in targets:
+            preds = predictions["classification"]
+            target = targets["classification"]
+            pred_classes = preds[0].argmax(dim=1)  # [B]
+            target_classes = target.argmax(dim=1)  # [B]
+            self.classification_cm += self.compute_confusion_matrix(
```
Comment (klemen1999) [inline luxonis_train/attached_modules/metrics/confusion_matrix.py:171]: Would it be better to have self.classification_cm be an actual torchmetric instance and then here in update we call update on it and in compute call compute on it? (same for segmentation_cm)
And since only one of them can be active (since only one task is allowed) can we get away with only one instance for both?

**Review by kozlov721 — APPROVED** [review]
